src/Stack/SDist.hs

```diff
@@ -465,51 +505,69 @@ buildExtractedTarball pkgDir = do
         sm {smProject = Map.insert (cpName $ ppCommon pp) pp pathsToKeep}
   local adjustEnvForBuild $ build Nothing Nothing
 
--- | Version of 'checkSDistTarball' that first saves lazy bytestring to
--- temporary directory and then calls 'checkSDistTarball' on it.
-checkSDistTarball'
+-- | Version of 'checkSDistTarball' that validates the tarball, which is available as a stream of
+-- bytes, rather than as a file on the file system. Contents will be extracted to a temporary
+-- directory for analysis, which must be supplied as an argument. Consumes GZipped tarball contents
+-- as a stream of bytes and collects it all into memory as lazy ByteString
+checkSDistTarballSink
   :: HasEnvConfig env
   => SDistOpts
-  -> String       -- ^ Tarball name
-  -> L.ByteString -- ^ Tarball contents as a byte string
-  -> RIO env ()
-checkSDistTarball' opts name bytes = withSystemTempDir "stack" $ \tpath -> do
-    npath   <- (tpath </>) `liftM` parseRelFile name
-    liftIO $ L.writeFile (toFilePath npath) bytes
-    checkSDistTarball opts npath
+  -> Path Abs Dir -- ^ Temporary path where tarball should be extracted to.
+  -> ConduitM ByteString Void (ResourceT (RIO env)) L.ByteString
+checkSDistTarballSink opts tpath = do
+    lbs <- C.getZipSink (C.ZipSink (extractTarGz tpath) *> C.ZipSink C.sinkLazy)
```

Comment:
`checkSDistTarballSink` produces lazy ByteString, maybe instead stream from tmp file, thus avoiding loading contents of the full tarball in memory